component/core.py

```diff
@@ -410,13 +450,27 @@ def component(self, usage=None, model_name=None):
                     component_classes,
                 )
             )
-        if model_name == self.model_name:
-            work_context = self
-        else:
-            work_context = self.work_on(model_name)
         return component_classes[0](work_context)
 
-    def many_components(self, usage=None, model_name=None):
+    def query_component(self, usage=None, model_name=None, **kw):
+        """Search for the 1st component matching given parameters.
+
+        Similar to `component` but does not break if no component it's found.
+        See `component` docstring for more details.
+        """
+        model_name = self._ensure_model_name(model_name)
+        component_classes, work_context = self._matching_components(
+            usage=usage, model_name=model_name, **kw
+        )
+        if len(component_classes) > 1:
+            component_classes = self._filter_components_by_collection(component_classes)
+        if len(component_classes) > 1:
+            component_classes = self._filter_components_by_model(
+                component_classes, model_name
+            )
+        return component_classes[0](work_context) if component_classes else None
```

Comment:
Thanks for the changes. The interrogation I had on the previous version remains. If more than one component match, the result is not deterministic (I assume, maybe you checked this). Then, as a user of the API, I don't know which component will be returned. How can it ever make sense?

Reply:
well, for instance, you may try to lookup a component by a sequence of usage candidates or by a sequence of other parameters. The order in this case does not matter, especially because normally you expect a component to be there or not. If it's not there you try w/ other parameters or simply you don't serve that feature for that given context/model/case.
Doing this w/out having to handle every time a try/except w/ relative import of the exception etc is munch more handy and clean, especially if you want to do this in many places and you don't won't to define your own api/util to handle this in the same way everywhere.
Hope you get my point :)

Reply:
> The order in this case does not matter, especially because normally you expect a component to be there or not.

You expect A component.
Ok, but you mention a case where you have one component or zero.

But again, if several match? You say the order does matter. I don't get why. If the list of matching component is > 1 and you return [0], and there is now way to control what will be zero, you may have a component and after a restart another component, that just feels wrong.

Reply:
And I think it IS wrong

Reply:
Well, yes, in theory it would be better to have a way to specify the ordering of the components.
Eg: if I have 2 params to match and I want to give precedence w/ this order:
1. param1 + param2 match
2. param1 only match
3. param2 only match

How would you handle this?
Right now there's no way to do it because there's no sorting and `_component_match` works on each single component class.

What you'll suggest -  I guess - is: use `many_components` sort resulting components and pick what you want. Right?

Reply:
> What you'll suggest - I guess - is: use many_components sort resulting components and pick what you want. Right?

Yes, it gives control to the user of the API.